raft/src/main/java/org/apache/kafka/raft/EpochState.java

```diff
@@ -26,16 +26,18 @@ default Optional<LogOffsetMetadata> highWatermark() {
     }
 
     /**
-     * Decide whether to grant a vote to a candidate.
+     * Decide whether to grant a vote to a replica.
      *
      * It is the responsibility of the caller to invoke
      * {@link QuorumState#transitionToUnattachedVotedState(int, ReplicaKey)} if vote is granted.
      *
-     * @param candidateKey the id and directory of the candidate
-     * @param isLogUpToDate whether the candidate’s log is at least as up-to-date as receiver’s log
+     * @param replicaKey the id and directory of the replica requesting the vote
+     * @param isLogUpToDate whether the replica's log is at least as up-to-date as receiver’s log
      * @return true if it can grant the vote, false otherwise
      */
-    boolean canGrantVote(ReplicaKey candidateKey, boolean isLogUpToDate);
+    boolean canGrantVote(ReplicaKey replicaKey, boolean isLogUpToDate);
+
+    boolean canGrantPreVote(ReplicaKey replicaKey, boolean isLogUpToDate);
```

Comment:
yes, originally it was just the one method since the logic for granting preVotes used to be (incorrectly) very similar to granting standard votes. I decided to split it out since I was seeing the cyclomatic complexity of certain states getting worse (e.g. Follower, Unattached state). 

I also considered having just one overridden epochState method `canGrantVote()` and then having that call a private helper `canGrantPreVote`, but that just seemed like an extra hoop

Reply:
> We do have this method call. It is in QuorumState::canGrantVote. If we didn't have this method, you would have the same if statement at every call site.

I agree with this statement, but not sure if I'm understanding if you are trying to make a further point.

> The other concern I have is which is easier to unittest? For example, it doesn't look like we have a unittest for QuorumState::canGrantVote.

As you said, QuorumState::canGrantVote just calls either EpochState::canGrantVote or EpochState::canGrantPreVote and both of those methods are unit tested?